### DISCUSSION TITLE
What deep learning training frameworks are supported?

### DISCUSSION THREAD

**Discussion by hw-ju** [Project-MONAI/monai-deploy-app-sdk#426] [Q&A]
Hi!
Can we use MONAI Deploy App SDK to package a model trained by any deep learning training frameworks, e.g., MONAI, PyTorch, Tensorflow, etc?  Or is there any requirement?

**Comment by MMelQin** [comment]
Thanks for the question @hw-ju. 

A quick answer is "Yes, any framework", though a long answer is needed to explain what is supported by the built-in functionalities/classes, and what will requires a bit of custom code.
- PyTorch models, specifically TorchScript, is supported by the built-in Model (loader) Factory, for simple as well as multiple model(s) (in a defined folder structure). Please see the [example](https://github.com/Project-MONAI/monai-deploy-app-sdk/tree/main/examples/apps/ai_multi_ai_app).
- For model trained without using MONAI, please see how the app can be easily built (model loading included) in this [example](https://github.com/Project-MONAI/monai-deploy-app-sdk/blob/main/examples/apps/mednist_classifier_monaideploy/mednist_classifier_monaideploy.py).
- Furthermore, for models trained with framework other than PyTorch, e.g. TF, an inference operator needs to be created to handle the model network loading, pass the pre-transformed tensors, and apply post transforms on the returned tensors. Also, the model file needs to be made available to the application at development time, as well as when the app is run as a container. One can make the model file a static content of the app, or rely on the MONAI App Package defined env var and container volume mappings for more dynamic model file access.
 
An example to demonstrate building an app for model(s) trained with other framework is not present yet in the repo, but is planned.

Hope this answers your question.

**Reply by hw-ju** [reply]
@MMelQin This is a great answer! Thank you so much!
